www/content/essays/web-security-basics-with-htmx.md

```diff
@@ -0,0 +1,322 @@
+
+Keep in mind that the JSON you parse might have a *property* that is formatted as HTML, though:
+
+```json
+{ "name": "<script>alert('Hahaha I am a script')</script>" }
+```
+
+Therefore, don't insert JSON values as HTML either—use `innerText` if you're doing something like that. This is well outside the realm of htmx-controlled UI though.
+
+The 2.0 version of htmx will include an `innerText` swap, if you want to call someone else's API directly from the client and just put that text into the page.
+
+### Custom HTML controls
+
+Unlike calling untrusted HTML routes, there are a lot of good reasons to let users do dynamic HTML-formatted content.
+
+What if, say, you want to let users link to an image?
+
+```html
+<img src="{{ user.fav_img }}">
+```
+
+Or link to their personal website?
+```html
+<a href="{{ user.fav_link }}">
+```
+
+The default "escape everything" approach escapes forward slashes, so it will bork user-submitted URLs.
+
+You can fix this in a couple of ways. The simplest, and safest, trick is to let users customize these values, but don't let them define the literal text. In the image example, you might upload the image to your own server (or S3 bucket, or the like), generate the link yourself, and then include it, unescaped. In nunjucks, you use the [safe](https://mozilla.github.io/nunjucks/templating.html#safe) function:
+
+```html
+<img src="{{ user.fav_img_s3_url | safe }}">
+```
+
+Yes, you're including unescaped content, but it's a link that you generated, so you know it's safe.
+
+You can handle custom CSS in the same way. Rather than let your users specific the color directly, give them some limited choices, and set the choices based on their input.
+
+```css
+{% if user.favorite_color === 'red' %}
+h1 { color: 'red'; }
+{% else %}
+h1 { color: 'blue'; }
+{% endif %}
+```
+
+In that example, the user can set `favorite_color` to whatever they like, but it's never going to be anything but red or blue. A less trivial example might ensure that only properly-formatted hex codes can be entered, using a regex. You get the idea.
+
+Depending on what kind of customization you're supporting, securing it might be relatively easy, or quite difficult. Some attributes are ["safe sinks,"](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#safe-sinks) which means that their values will never be interpreted as code; these are quite easy to secure. If you're going to include dynamic input in ["dangerous contexts,"](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#dangerous-contexts) you need to research *what* is dangerous about those contexts, and ensure that that kind of input won't make it into the document.
```

Comment:
Same as before, recommend moving the commas to the right of the link markup ie `[...](),`